feat: convert unpinned Ubuntu EC2NodeClasses (#6699)
aws/karpenter-provider-aws: website/content/en/preview/upgrading/v1-migration.md

* **metadataOptions could break workloads**: If you have workload pods that are not using `hostNetworking`, the updated default `metadataOptions` could cause your containers to break when you apply new EC2NodeClasses on v1.

* **Ubuntu AMIFamily Removed**:

   Support for automatic AMI selection and UserData generation for Ubuntu has been dropped with Karpenter `v1.0.0`.
   To continue using Ubuntu AMIs you will need to specify an AMI using `amiSelectorTerms`.

>    UserData generation can be achieved using the AL2 AMIFamily which, at the time of writing, has an identical UserData format.

Reviewer: It's implied that at the time of writing it was identical. I think we can remove. It's unsupported anyway, so user will have to validate it themselves anyway.
```suggestion
   UserData generation can be achieved using the AL2 AMIFamily which has an identical UserData format.
```

>    However, compatibility is not guaranteed long-term and changes to either AL2 or Ubuntu's UserData format my introduce incompatibilities.

Reviewer 2: ```suggestion
   However, compatibility is not guaranteed long-term and changes to either AL2 or Ubuntu's UserData format may introduce incompatibilities.
```

   If this occurs, the Custom AMIFamily should be used for Ubuntu and UserData will need to be entirely maintained by the user.

   If you are upgrading to `v1.0.0` and already have v1beta1 Ubuntu EC2NodeClasses, all you need to do is specify `amiSelectorTerms` and Karpenter will translate your NodeClasses to the v1 equivalent (as shown below).

>    Failure to specify `amiSelectorTerms` will result in the EC2NodeClass and all associated NodePools going unready.

Reviewer: ```suggestion
   Failure to specify `amiSelectorTerms` will result in the EC2NodeClass and all referencing NodePools to show as NotReady, causing Karpenter to ignore these NodePools and EC2NodeClasses for Provisioning and Drift.
```

### Downgrading
// TODO

## Full Changelog
* Features:
  * AMI Selector Terms has a new Alias field which can only be set by itself in `EC2NodeClass.Spec.AMISelectorTerms`
  * Disruption Budgets by Reason was added to `NodePool.Spec.Disruption.Budgets`
  * TerminationGracePeriod was added to `NodePool.Spec.Template.Spec`.
  * LOG_OUTPUT_PATHS and LOG_ERROR_OUTPUT_PATHS environment variables added
* API Rename: NodePool’s ConsolidationPolicy `WhenUnderutilized` is now renamed to `WhenEmptyOrUnderutilized`
* Behavior Changes:
  * Expiration is now forceful and begins draining as soon as it’s expired. Karpenter does not wait for replacement capacity to be available before draining, but will start provisioning a replacement as soon as the node is expired and begins draining.
  * Karpenter's generated NodeConfig now takes precedence when generating UserData with the AL2023 `amiFamily`. If you're setting any values managed by Karpenter in your AL2023 UserData, configure these through Karpenter natively (e.g. kubelet configuration fields).

>   * Karpenter now adds a `karpenter.sh/unregistered` taint to nodes in injected UserData when using alias in AMISelectorTerms or non-Custom AMIFamily. When using `amiFamily: Custom`, users will need to add this taint into their UserData, where Karpenter will automatically remove it when provisioning nodes.

Reviewer: Let's add the NoExecute like you did above.
```suggestion
  * Karpenter now adds a `karpenter.sh/unregistered:NoExecute` taint to nodes in injected UserData when using alias in AMISelectorTerms or non-Custom AMIFamily. When using `amiFamily: Custom`, users will need to add this taint into their UserData, where Karpenter will automatically remove it when provisioning nodes.
```
